File: PackIt/src/kotlin/sh/packit/core/utils/Paths.kt
package sh.packit.core.utils

import android.content.Context
import de.shareui.exterasdk.cruel.PluginPaths
import org.telegram.messenger.ApplicationLoader
import sh.packit.core.state.CoreState
import java.io.File

object Paths {
    private fun getContext(): Context? {
        return try {
            ApplicationLoader.applicationContext
        } catch (e: Throwable) {
            Logx.logx("failed to get ApplicationLoader context: $e", isDebug = false)
            null
        }
    }

    fun getFilesDir(): File? = getContext()?.filesDir

    fun getPluginsDir(): File? {
        val filesDir: File = getFilesDir() ?: return null
        return File(filesDir, "plugins")
    }

    fun getPluginSettingsFile(): File? {
        val pluginsDir: File = getPluginsDir() ?: return null
        return File(pluginsDir, "plugin_settings.json")
    }

    fun getPackitDir(): File? {
        val filesDir: File = getFilesDir() ?: return null
        return File(filesDir, "packit")
    }

    fun getPackitVarDir(): File? {
        val packitDir: File = getPackitDir() ?: return null
        return File(packitDir, "var")
    }

    fun getLogsDir(): File? {
        val varDir: File = getPackitVarDir() ?: return null
        return File(varDir, "logs")
    }

    fun getHistoryDir(): File? {
        val logsDir: File = getLogsDir() ?: return null
        return File(logsDir, "history")
    }

    fun getLatestLogFile(): File? {
        val logsDir: File = getLogsDir() ?: return null
        return File(logsDir, "latest.txt")
    }

    fun getAssetsDir(pluginId: String = CoreState.PLUGIN_ID, explicitAssetsDir: String? = null): File? {
        if (!explicitAssetsDir.isNullOrEmpty()) {
            return File(explicitAssetsDir)
        }
        val stateDir: String? = CoreState.assetsDir
        if (!stateDir.isNullOrEmpty()) {
            return File(stateDir)
        }
        return try {
            PluginPaths.getAssetsDir(pluginId)
        } catch (e: Throwable) {
            Logx.logx("failed to resolve assets dir: $e", isDebug = false)
            null
        }
    }

    fun getFontsDir(pluginId: String = CoreState.PLUGIN_ID, explicitAssetsDir: String? = null): File? {
        val assets: File = getAssetsDir(pluginId, explicitAssetsDir) ?: return null
        return if (assets.name == "fonts") assets else File(assets, "fonts")
    }
}

File: PackIt/src/wheels/packutil/package/paths.py
# pyright: reportMissingImports=false

from pathlib import Path
from cruel import get_assets_dir, get_locale_dir, get_share_dir, get_source_dir
from .logx import logx

def getFilesDir() -> str:
    try:
        from org.telegram.messenger import ApplicationLoader
        return ApplicationLoader.applicationContext.getFilesDir().getAbsolutePath()
    except Exception as e:
        logx(f"failed to get filesDir: {e}", isDebug=False)
        raise

def getCacheDir() -> str:
    try:
        from org.telegram.messenger import ApplicationLoader
        return ApplicationLoader.applicationContext.getCacheDir().getAbsolutePath()
    except Exception as e:
        logx(f"failed to get cacheDir: {e}", isDebug=False)
        raise

def getPluginsDir() -> str:
    return getFilesDir() + "/plugins"

def getPluginSettingsPath() -> str:
    return getPluginsDir() + "/plugin_settings.json"

def getPackitDir() -> str:
    return getFilesDir() + "/packit"

def getPackitVarDir() -> str:
    return getPackitDir() + "/var"

def getLogsDir() -> str:
    return getPackitVarDir() + "/logs"

def getHistoryDir() -> str:
    return getLogsDir() + "/history"

def getLatestLogPath() -> str:
    return getLogsDir() + "/latest.txt"

def getConfigsDir() -> str:
    return getPackitDir() + "/packitConfigs"

def getReposCacheDir() -> str:
    return getPackitDir() + "/reposCache"

def getTempDir() -> str:
    return getPackitDir() + "/packitTemp"

def getAssetsDir(pluginId: str = "packit") -> Path:
    try:
        return get_assets_dir(pluginId)
    except Exception as e:
        logx(f"failed to get assetsDir for {pluginId}: {e}", isDebug=False)
        raise

def getSourceDir(pluginId: str = "packit") -> Path:
    try:
        return get_source_dir(pluginId)
    except Exception as e:
        logx(f"failed to get sourceDir for {pluginId}: {e}", isDebug=False)
        raise

def getShareDir(pluginId: str = "packit") -> Path:
    try:
        return get_share_dir(pluginId)
    except Exception as e:
        logx(f"failed to get shareDir for {pluginId}: {e}", isDebug=False)
        raise

def getLocaleDir(pluginId: str = "packit") -> Path:
    try:
        return get_locale_dir(pluginId)
    except Exception as e:
        logx(f"failed to get localeDir for {pluginId}: {e}", isDebug=False)
        raise

def getDexPath(name: str = "Core.dex", pluginId: str = "packit") -> Path:
    return getAssetsDir(pluginId) / name

def getCoreDexPath(pluginId: str = "packit") -> Path:
    return getDexPath("Core.dex", pluginId)

def getComposeDexPath(pluginId: str = "packit") -> Path:
    return getDexPath("Compose.dex", pluginId)
